src/ape/api/providers.py

```diff
@@ -1591,6 +1625,283 @@ def _handle_execution_reverted(
         )
         return self.compiler_manager.enrich_error(result)
 
+    @property
+    def head(self) -> BlockAPI:
+        return self.get_block("latest")
+
+    @property
+    def height(self) -> int:
+        return self.head.number or 0
+
+    def range_blocks(
+        self,
+        start_or_stop: int,
+        stop: Optional[int] = None,
+        step: int = 1,
+        engine_to_use: Optional[str] = None,
+    ) -> Iterator[BlockAPI]:
+        """
+        Iterate over blocks. Works similarly to python ``range()``.
+
+        Raises:
+            :class:`~ape.exceptions.ChainError`: When ``stop`` is greater
+                than the chain length.
+            :class:`~ape.exceptions.ChainError`: When ``stop`` is less
+                than ``start_block``.
+            :class:`~ape.exceptions.ChainError`: When ``stop`` is less
+                than 0.
+            :class:`~ape.exceptions.ChainError`: When ``start`` is less
+                than 0.
+
+        Args:
+            start_or_stop (int): When given just a single value, it is the stop.
+              Otherwise, it is the start. This mimics the behavior of ``range``
+              built-in Python function.
+            stop (Optional[int]): The block number to stop before. Also the total
+              number of blocks to get. If not setting a start value, is set by
+              the first argument.
+            step (Optional[int]): The value to increment by. Defaults to ``1``.
+             number of blocks to get. Defaults to the latest block.
+            engine_to_use (Optional[str]): query engine to use, bypasses query
+              engine selection algorithm.
+
+        Returns:
+            Iterator[:class:`~ape.api.providers.BlockAPI`]
+        """
+
+        if stop is None:
+            stop = start_or_stop
+            start = 0
+        else:
+            start = start_or_stop
+
+        if stop > len(self):
+            raise ProviderError(
+                f"'stop={stop}' cannot be greater than the chain length ({len(self)}). "
+                f"Use '{self.poll_blocks.__name__}()' to wait for future blocks."
+            )
+
+        # Note: the range `stop_block` is a non-inclusive stop, while the
+        #       `.query` method uses an inclusive stop, so we must adjust downwards.
+        query = BlockQuery(
+            columns=list(self.head.__fields__),  # TODO: fetch the block fields from EcosystemAPI
+            start_block=start,
+            stop_block=stop - 1,
+            step=step,
+        )
+
+        blocks = self.query_manager.query(query, engine_to_use=engine_to_use)
+        yield from cast(Iterator[BlockAPI], blocks)
+
+    def poll_blocks(
```

Comment:
Technically, for web3py at least, there's a better implementation of this using `eth_newBlockFilter`:
https://web3py.readthedocs.io/en/stable/web3.eth.html#web3.eth.Eth.filter